service-base/src/main/java/org/eclipse/hono/service/amqp/RequestResponseEndpoint.java

```diff
@@ -237,54 +296,179 @@ protected final void handleMessage(final ProtonConnection con, final ProtonRecei
+                    sender.getName(), replyConsumer.address());
+
+            deallocateReceiverForReplyTo(replyToAddress.toString());
+            unregisterConsumerForConnection(con, replyTo, replyConsumer);
+
+            if (senderClosed.succeeded()) {
+                senderClosed.result().close();
+            }
+        });
+
+        sender.open();
+
+    }
+
+    @Override
+    public void onConnectionClosed(final ProtonConnection connection) {
+        unregisterAllConsumersForConnection(connection);
+        deallocateAllReceiversForConnection(connection);
+    }
+
+    private void allocateReceiverForReplyTo(final String replyTo, final ProtonReceiver receiver) {
+
+        final ProtonReceiver oldReceiver = replyToReceiverMap.put(replyTo, receiver);
+
+        if (oldReceiver == null || oldReceiver == receiver) {
+            logger.debug("Allocated receiver [{}] for replies to [{}]", receiver, replyTo);
+        } else {
+            logger.info("Allocated receiver [{}] for replies to [{}] - Had existing receiver: [{}]", receiver, replyTo,
+                    oldReceiver);
+        }
+    }
+
+    private void flowCreditToRequestor(final String replyTo) {
+
+        final ProtonReceiver receiver = replyToReceiverMap.get(replyTo);
+        if (receiver == null) {
+            logger.warn("No receiver found for reply-to address", replyTo);
+            return;
+        }
+
+        // flow one credit back to the receiver
+        receiver.flow(1);
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("Flowing credit back to sender - replyTo: [{}], currentCredits: {}", replyTo,
+                    receiver.getCredit());
+        }
+
+    }
+
+    private void deallocateReceiverForReplyTo(final String replyTo) {
+
+        final ProtonReceiver receiver = replyToReceiverMap.remove(replyTo);
+        if (receiver == null) {
+            logger.warn("Receiver [{}] was not allocated to replyTo address [{}]", replyTo);
+        } else {
+            logger.debug("Deallocated receiver [{}] for replies to [{}]", receiver, replyTo);
         }
+
+    }
+
+    private void deallocateAllReceiversForConnection(final ProtonConnection connection) {
+        replyToReceiverMap
+                .entrySet()
+                .removeIf(entry -> entry.getValue().getSession().getConnection() == connection);
+    }
+
+    private void registerConsumerForConnection(final ProtonConnection connection,
+            final String replyTo, final MessageConsumer<?> replyConsumer) {
+
+        replyConsumerMap.put(connection, replyConsumer);
+        replyAddressMap.put(connection, replyTo);
+        replyAddresses.add(replyTo);
+
+    }
+
+    private void unregisterConsumerForConnection(final ProtonConnection connection,
+            final String replyTo, final MessageConsumer<?> replyConsumer) {
+
+        // unregister the consumer
+
+        replyConsumer.unregister();
```

Comment:
AFAIK the `org.eclipse.hono.client.MessageConsumer` interface has no such method, or am I mistaken?

Reply:
That may be, but that is a `io.vertx.core.eventbus.MessageConsumer`, and it does compile.